common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -5846,6 +5659,77 @@ static unsigned int psk_server_callback(SSL* ssl, const char *identity,
     return static_cast<unsigned int>(keyLen);
 }
 
+static int new_session_callback(SSL* ssl, SSL_SESSION* session) {
+    JNI_TRACE("ssl=%p new_session_callback session=%p", ssl, session);
+
+    AppData* appData = toAppData(ssl);
+    JNIEnv* env = appData->env;
+    if (env == nullptr) {
+        ALOGE("AppData->env missing in new_session_callback");
+        JNI_TRACE("ssl=%p new_session_callback env error", ssl);
+        return 0;
+    }
+    if (env->ExceptionCheck()) {
+        JNI_TRACE("ssl=%p new_session_callback already pending exception", ssl);
+        return 0;
+    }
+
+    jobject sslHandshakeCallbacks = appData->sslHandshakeCallbacks;
+    jclass cls = env->GetObjectClass(sslHandshakeCallbacks);
+    jmethodID methodID = env->GetMethodID(cls, "onNewSessionEstablished", "(J)Z");
+    JNI_TRACE("ssl=%p new_session_callback calling onNewSessionEstablished", ssl);
+    jboolean ret = env->CallBooleanMethod(sslHandshakeCallbacks, methodID,
+                                          reinterpret_cast<jlong>(session));
+    if (env->ExceptionCheck()) {
+        JNI_TRACE("ssl=%p new_session_callback exception cleared", ssl);
+        env->ExceptionClear();
+    }
+    JNI_TRACE("ssl=%p new_session_callback completed => %d", ssl, ret);
+    return ret;
+}
+
+static SSL_SESSION* server_session_requested_callback(SSL* ssl, uint8_t* id, int id_len,
+                                                      int* out_copy) {
+    JNI_TRACE("ssl=%p server_session_requested_callback", ssl);
+
+    // Always set to out_copy to zero. The Java callback will be responsible for incrementing
+    // the reference count (and any required synchronization).
+    *out_copy = 0;
+
+    AppData* appData = toAppData(ssl);
+    JNIEnv* env = appData->env;
+    if (env == nullptr) {
+        ALOGE("AppData->env missing in server_session_requested_callback");
+        JNI_TRACE("ssl=%p server_session_requested_callback env error", ssl);
+        return 0;
+    }
+    if (env->ExceptionCheck()) {
+        JNI_TRACE("ssl=%p server_session_requested_callback already pending exception", ssl);
+        return 0;
+    }
+
+    // Copy the ID to a byte[].
+    jbyteArray id_array = env->NewByteArray(static_cast<jsize>(id_len));
+    if (id_array == nullptr) {
+        JNI_TRACE("ssl=%p id_array bytes == null => 0", ssl);
+        return 0;
+    }
+    env->SetByteArrayRegion(id_array, 0, static_cast<jsize>(id_len),
+                            reinterpret_cast<const jbyte*>(id));
+
+    jobject sslHandshakeCallbacks = appData->sslHandshakeCallbacks;
+    jclass cls = env->GetObjectClass(sslHandshakeCallbacks);
+    jmethodID methodID = env->GetMethodID(cls, "serverSessionRequested", "([B)J");
+    JNI_TRACE("ssl=%p server_session_requested_callback calling serverSessionRequested", ssl);
+    jlong ssl_session_address = env->CallLongMethod(sslHandshakeCallbacks, methodID, id_array);
```

Comment:
The lifetimes here are very non-obvious without C and Java halves together. See the mess with the new session callback. The thing about these callbacks across JNI is that you cannot review one without the other. I believe this is correct, but you need to remember to `SSL_SESSION_up_ref` in Java *while the SslSessionWrapper is alive*.

It would be much better to take it out so that, when the Java half is reviewed, the C half is in the diff at the same time for simultaneous review.